### PR TITLE
Add FromText instance to S3Uri

### PR DESCRIPTION
**Pull request by newhoggy**

### CODE REVIEW

File: antiope-s3/test/Antiope/S3/TypesSpec.hs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+spec = describe "Antiope.TypesSpec" $ do
+  it "Can parse S3Uri" $ requireTest $ do
+    fromText ""                 === (Left "not enough input"  :: Either String S3Uri)
+    fromText "boo"              === (Left "string"            :: Either String S3Uri)
+    fromText "s3://hello"       === Right (S3Uri (BucketName "hello") (ObjectKey ""))
+    fromText "s3://hello/"      === Right (S3Uri (BucketName "hello") (ObjectKey ""))
+    fromText "s3://hello/a"     === Right (S3Uri (BucketName "hello") (ObjectKey "a"))
+    fromText "s3://hello/ab"    === Right (S3Uri (BucketName "hello") (ObjectKey "ab"))
+    fromText "s3://hello/ab "   === Right (S3Uri (BucketName "hello") (ObjectKey "ab"))
+    fromText "s3://hello/ab c"  === Right (S3Uri (BucketName "hello") (ObjectKey "ab"))
```
Comment (AlexeyRaga) [inline antiope-s3/test/Antiope/S3/TypesSpec.hs]: Why this would ever be the case?

File: antiope-s3/src/Antiope/S3/Types.hs [hunk 2]
```diff
@@ -33,6 +36,14 @@ data S3Uri = S3Uri
   , objectKey :: ObjectKey
   } deriving (Show, Eq, Generic)
 
+instance FromText S3Uri where
+  parser = do
+    _  <- DAT.string "s3://"
+    bn <- BucketName . T.pack <$> DAC.many1 (DAT.satisfy (\c -> c /= '/' && c /= ' '))
```
Comment (Kheldar) [inline antiope-s3/src/Antiope/S3/Types.hs]: Perhaps change the `/= ' '` to some sort of generic doesn't-match-whitespace?

Comment (newhoggy) [reply antiope-s3/src/Antiope/S3/Types.hs]: Turns out space is valid.  Does that mean s3 uris aren't really uris?

Comment (AlexeyRaga) [reply antiope-s3/src/Antiope/S3/Types.hs]: I think it can follow the hadoop FS specs... But I am not sure

Comment (AlexeyRaga) [reply antiope-s3/src/Antiope/S3/Types.hs]: https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html

Comment (AlexeyRaga) [reply antiope-s3/src/Antiope/S3/Types.hs]: Space is listed as " might require additional code handling and likely will need to be URL encoded or referenced as HEX"

Comment (newhoggy) [reply antiope-s3/src/Antiope/S3/Types.hs]: I think it's amazonka's responsibility to encode it in that case.

File: antiope-s3/test/Antiope/S3/TypesSpec.hs (hunk 1 above)
Comment (newhoggy) [reply antiope-s3/test/Antiope/S3/TypesSpec.hs]: I've fixed the code.  Please review.

File: antiope-s3/src/Antiope/S3/Types.hs (hunk 2 above)
Comment (AlexeyRaga) [reply antiope-s3/src/Antiope/S3/Types.hs]: If Amazonka expects "raw" text as the parameter to `ObjectKey` then sure.

**Pull request merged**